src/cdomains/apron/linearTwoVarEqualityDomain.apron.ml

```diff
@@ -0,0 +1,779 @@
+
+  let assign_var_parallel_with t vv's =
+    let t' = assign_var_parallel t vv's in
+    t.d <- t'.d;
+    t.env <- t'.env
+
+  let assign_var_parallel_with t vv's =
+    if M.tracing then M.tracel "var_parallel" "assign_var parallel'\n";
+    assign_var_parallel_with t vv's
+
+  let assign_var_parallel' t vs1 vs2 =
+    let vv's = List.combine vs1 vs2 in
+    assign_var_parallel t vv's
+
+  let assign_var_parallel' t vv's =
+    let res = assign_var_parallel' t vv's in
+    if M.tracing then M.tracel "ops" "assign_var parallel'\n";
+    res
+
+  let substitute_exp ask t var exp no_ov =
+    let t = if not @@ Environment.mem_var t.env var then add_vars t [var] else t in
+    let res = assign_exp ask t var exp no_ov in
+    forget_vars res [var]
+
+  let substitute_exp ask t var exp no_ov =
+    let res = substitute_exp ask t var exp no_ov
+    in if M.tracing then M.tracel "ops" "Substitute_expr t: \n %s \n var: %s \n exp: %a \n -> \n %s\n" (show t) (Var.to_string var) d_exp exp (show res);
+    res
+
+  let substitute_exp ask t var exp no_ov = timing_wrap "substitution" (substitute_exp ask t var exp) no_ov
+
+  let show_coeff_vec l (env : Environment.t) =
+    let show_element e = match e with
+      | (a, Some x) -> ((Z.to_string a) ^ " * " ^ (Var.to_string ( Environment.var_of_dim env x)) ^ " + ")
+      | (a, None) -> ((Z.to_string a) ^ "+") in
+    List.fold_right (fun k result -> show_element k ^ "\n" ^ result) l ""
+
+  let show_final_expr l (env : Environment.t) =
+    let show_element i a = if i = 0 then ((Z.to_string a) ^ " + ") else
+        ((Z.to_string a) ^ " * " ^ (Var.to_string ( Environment.var_of_dim env (i-1))) ^ " + ") in
+    List.fold_righti (fun i k result -> show_element i k ^ "\n" ^ result) l ""
+
+  (** Assert a constraint expression.
+      The overflow is completely handled by the flag "no_ov",
+      which is set in relationAnalysis.ml via the function no_overflow.
+      In case of a potential overflow, "no_ov" is set to false
+      and Convert.tcons1_of_cil_exp will raise the exception Unsupported_CilExp Overflow *)
+
+  (* meet_tcons -> meet with guard in if statement
+     texpr -> tree expr (right hand side of equality)
+     -> expression used to derive tcons -> used to check for overflow
+     tcons -> tree constraint (expression < 0)
+     -> does not have types (overflow is type dependent)
+  *)
+
+  let meet_tcons ask t tcons original_expr no_ov =
+    (* The expression is evaluated using an array of coefficients. The first element of the array belongs to the constant followed by the coefficients of all variables
+       depending on the result in the array after the evaluating including resolving the constraints in t.d the tcons can be evaluated and additional constraints can be added to t.d *)
+    match t.d with
+    | None -> bot_env
+    | Some d ->
+      let expr = Array.init (Environment.size t.env) (fun _ -> Z.zero) in
+      let constant = ref (Z.zero) in
```

Comment:
The following things all look quite involved, did you write regression tests for the more complicated cases here?

Reply:
I added another test which includes 4 variables with two different dependencies which are non constant and a constant. It works. (Test 33)